### PR TITLE
Opal-207 backend can process account removal

### PR DESCRIPTION
**Pull request by sznemeth3**

### CODE REVIEW

**Review by markaron — DISMISSED** [review]

**Review by pataricza1987 — DISMISSED** [review]

File: src/main/java/com/greenfoxacademy/opal/kalendaryo/kalendaryo/controllers/AuthController.java [hunk 1]
```diff
@@ -18,45 +20,22 @@
...
-                userResponse.getGoogleAuths().add(currentGetAuthResponse);
-            }
-
-            return ResponseEntity.status(200).body(new UserResponse(kalUser.getId(), kalUser.getUserEmail(), userResponse.getGoogleAuths()));
-        }
-        return ResponseEntity.status(401).body("Client token is missing or invalid");
+  @PostMapping("/auth")
+  public ResponseEntity postAuth(@RequestBody GoogleAuth googleAuth, @RequestHeader("X-Client-Token") String clientToken, HttpServletRequest request) throws IOException {
+    try {
+      return ResponseEntity.status(200).body(authAndUserService.createPostAuthResponse(clientToken, googleAuth));
```
Comment (arnikaeva) [inline src/main/java/com/greenfoxacademy/opal/kalendaryo/kalendaryo/controllers/AuthController.java]: you can use 2 lines, so the `authAndUserService.createPostAuthResponse(clientToken, googleAuth)` call is not inside the body()

File: src/main/java/com/greenfoxacademy/opal/kalendaryo/kalendaryo/service/AuthAndUserService.java [hunk 2]
```diff
@@ -11,49 +14,128 @@
...
+    try {
+      KalUser user = findUserByClientToken(clientToken);
+      if (!checkIfEmailIsLoggedInUser(email, user)) {
+        GoogleAuth googleAuth = googleAuthRepository.findByEmailAndUser_Id(email, user.getId());
+        googleAuthRepository.delete(googleAuth);
+      } else {
+        throw new ValidationException("Primary email can not be deleted!");
+      }
+    } catch (NullPointerException e) {
+      throw new ValidationException("Client token/email is missing or invalid");
```
Comment (arnikaeva) [inline src/main/java/com/greenfoxacademy/opal/kalendaryo/kalendaryo/service/AuthAndUserService.java]: would be nice to know the exact error :)

File: src/main/java/com/greenfoxacademy/opal/kalendaryo/kalendaryo/service/AuthAndUserService.java [hunk 3]
```diff
@@ -11,49 +14,128 @@
...
+      googleAuth = manageGoogleAuthForPostAuth(kalUser, googleAuth);
+      return new PostAuthResponse(kalUser.getId(), kalUser.getClientToken(), googleAuth.getEmail(), googleAuth.getAccessToken());
+    } catch (IOException e) {
+      throw new ValidationException("Error while saving");
     }
+  }
 
-    public KalUser findUserByAuth(GoogleAuth googleAuth) {
-        return kalUserRepository.findKalUserByGoogleAuthListIsContaining(googleAuth);
+  public KalUser setKalUserForPostAuth(String clientToken, GoogleAuth googleAuth) {
```
Comment (arnikaeva) [inline src/main/java/com/greenfoxacademy/opal/kalendaryo/kalendaryo/service/AuthAndUserService.java:122]: would refactor this method so it is evident what happens and why in the various cases

Comment (arnikaeva) [reply src/main/java/com/greenfoxacademy/opal/kalendaryo/kalendaryo/service/AuthAndUserService.java:122]: something like
```
if (clientExists(clientToken) {
..
} else if (someOtherCheckerMethod()) { //not sure what happens here
...
} else {
  kalUser =  createNewUser();
}
```

File: src/test/java/com/greenfoxacademy/opal/kalendaryo/kalendaryo/service/AuthAndUserServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  @Test
+  public void createPostAuthResponse_testIfResponseIsCreated() throws ValidationException{
+    String email = "test@email.com";
+    String authCode = "testCode";
+    String displayName = "Test Elek";
+    KalUser kalUser = new KalUser();
+    String accesToken = "testToken";
+    GoogleAuth googleAuth = new GoogleAuth(email, authCode, displayName, kalUser, accesToken);
```
Comment (arnikaeva) [inline src/test/java/com/greenfoxacademy/opal/kalendaryo/kalendaryo/service/AuthAndUserServiceTest.java]: if you do not use the variables later on (in an assert) than there is no need to declare them, can use here  `GoogleAuth googleAuth = new GoogleAuth("email", "etc", ....);`

File: src/main/java/com/greenfoxacademy/opal/kalendaryo/kalendaryo/service/AuthAndUserService.java [hunk 5]
```diff
@@ -11,49 +14,128 @@
...
+    if (checkIfGoogleAuthExist(googleAuth, kalUser.getId())) {
+      googleAuth = googleAuthRepository.findByEmailAndUser_Id(googleAuth.getEmail(), kalUser.getId());
+      return googleAuth;
+    }
+    saveGoogleAuth(googleAuth);
+    return googleAuth;
+  }
+
+  public boolean checkIfGoogleAuthExist(GoogleAuth googleAuth, Long kalUserId) {
+    return googleAuthRepository.findByEmailAndUser_Id(googleAuth.getEmail(), kalUserId) != null;
```
Comment (arnikaeva) [inline src/main/java/com/greenfoxacademy/opal/kalendaryo/kalendaryo/service/AuthAndUserService.java]: will this really be null? Or will this throw a nullPointerException?

**Review by arnikaeva — CHANGES_REQUESTED** [review]

**Review by arnikaeva — DISMISSED** [review]

**Review by pataricza1987 — APPROVED** [review]

File: src/main/java/com/greenfoxacademy/opal/kalendaryo/kalendaryo/controllers/AccountController.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      return ResponseEntity.status(400).body(val.getMessage());
+    }
+  }
+
+  @Transactional
+  @DeleteMapping(value = "/account")
+  public ResponseEntity deleteGoogleAccount(@RequestHeader("X-Client-Token") String clientToken, @RequestHeader("email") String email,
+                                            HttpServletRequest request) throws IOException {
+    try {
+      authAndUserService.deleteGoogleAuth(email, clientToken);
```
Comment (markaron) [inline src/main/java/com/greenfoxacademy/opal/kalendaryo/kalendaryo/controllers/AccountController.java:36]: Nice!

**Review by markaron — APPROVED** [review]

**Pull request merged**